### PR TITLE
SPM-58: Refactor Quiz Attempt

### PR DESCRIPTION
**Pull request by joelytj**
## Summary
Refactored Quiz Attempt for both graded and ungraded quizzes

- For both ungraded quiz and graded quiz, the score and the questions wrong will be in the output.
- For ungraded quiz, any attempt is considered ‘completed’, and a sectionCompleted record will be created. 
- For graded quiz, if the quiz_attempt passes, a sectionCompleted record will be created and the Enrollment record for the engineer will be updated to has_passed=True. Otherwise, it will only return the score and the questions wrong.

## Changes
1. Created `Answer` model and schema
2. Created `GET` and `POST` under `resources/quiz_attempt.py` (bulk of the PR)
3. Created tests under `tests/test_quiz_attempt.py`
4. All other files are unchanged

## Note
For Quiz, `is_graded default=False`, so when creating the **last** quiz (of the last section) it should be set to `True,` and a `passing_mark` should be given. (e.g passing_mark = 1 => minimally have to get 2 questions correct to pass the quiz)

Here are some sample outputs for `POST QuizAttempt`. @edwinlzs @xiulingchua Let me know if you guys need help with the input parameters/output for `Quiz/QuizAttempt` endpoints

Screenshot 2021-11-01 at 9 28 28 PM
---------------------------------------------------------------------------------------------------------
Screenshot 2021-11-01 at 10 03 50 PM
---------------------------------------------------------------------------------------------------------
Screenshot 2021-11-01 at 9 25 33 PM

### CODE REVIEW

**Comment by yarkhinephyo** [comment]
Not sure what happened but there is a lot of frontend changes. If you are having difficulties to resolve conflicts line by line, a clean and easy way to resolve the conflicts is:
1. Look at which frontend files are changed in this PR (For example - `frontend/.eslintrc.json`, `frontend/public/favicon.ico`, `frontend/public/index.html` and so on .... )
2. Git checkout `release4` branch and copy those files into local laptop
3. Git checkout `Refactor-Quiz-Attempt` and replace/ add those files with the ones that have been copied into local laptop
4. Commit and push

If things go well, in the `Files Changed` tab, you will be left with only backend changes

File: restful_api/myapi/api/resources/quiz_attempt.py [hunk 1]
```diff
@@ -48,14 +48,19 @@ class QuizAttemptResource(Resource):
...
     def get(self, course_id, section_id, trainer_id, eng_id):
         try:
             query = (
                 QuizAttempt.query
-                .join(Quiz, (QuizAttempt.section_id == Quiz.section_id) & (QuizAttempt.course_id == Quiz.course_id), isouter=True)
                 .join(Employee, isouter=True)
                 .filter(Employee.id == eng_id)
+                .filter(QuizAttempt.course_id == course_id)
+                .filter(QuizAttempt.section_id == section_id)
+                .filter(QuizAttempt.trainer_id == trainer_id)
```
Comment (yarkhinephyo) [inline restful_api/myapi/api/resources/quiz_attempt.py:63]:
```
.filter((QuizAttempt.course_id == course_id) & (QuizAttempt.section_id == section_id) & (QuizAttempt.trainer_id == trainer_id))
```

This should work but not important, up to u

File: restful_api/myapi/api/resources/quiz_attempt.py [hunk 2]
```diff
@@ -67,21 +72,142 @@ def get(self, course_id, section_id, trainer_id, eng_id):
...
+
+        # If quiz_attempt incomplete, set answer_labels for unattempted questions to be None
+        if len(answers) < len(fmted_questions):
+            for i in range(len(fmted_questions) - len(answers)):
+                answers.append({"answer_label": None})
+
+        wrong = []
+        num_correct = 0
+        for i in range(len(fmted_questions)):
+            if fmted_questions[i][1] == answers[i]['answer_label']:
```
Comment (yarkhinephyo) [inline restful_api/myapi/api/resources/quiz_attempt.py:203]: May at the start of for loop just add a sanity check to make sure the question and answers are the same.

```
assert fmted_questions[i][0] == answers[i]['question_id'], "Wrong question and answer matching"

```

File: restful_api/myapi/api/resources/quiz_attempt.py (hunk 1 above)
Comment (joelytj) [reply restful_api/myapi/api/resources/quiz_attempt.py:63]: oh ya HAHA forgot I could do it that way

File: restful_api/myapi/api/resources/quiz_attempt.py (hunk 2 above)
Comment (joelytj) [reply restful_api/myapi/api/resources/quiz_attempt.py:203]: ok!

File: restful_api/myapi/api/schemas/answer.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+from myapi.models import Answer
+from myapi.extensions import ma, db
+
+
+class AnswerSchema(ma.SQLAlchemyAutoSchema):
+
+    course_id = ma.Int(required=False)
+    trainer_id = ma.Int(required=False)
+    section_id = ma.Int(required=False)
+    eng_id = ma.Int(required=False)
```
Comment (yarkhinephyo) [inline restful_api/myapi/api/schemas/answer.py:10]:
```
    course_id = ma.Int(required=False, load_only=True)
    trainer_id = ma.Int(required=False, load_only=True)
    section_id = ma.Int(required=False, load_only=True)
    eng_id = ma.Int(required=False, load_only=True)
```

Might look neater for frontend if you dun dump those variables. Up to u!

**Review by yarkhinephyo — APPROVED** [review]
I think can wait for @edwinlzs PR to fix test then merge. Oh Python linter also got problem `QuizAttemptResource.post` function is too complex. Can break it down

**Review by yarkhinephyo — APPROVED** [review]

**Pull request merged**